### PR TITLE
Ensure that the src file contents is converted to unicode in diff info

### LINKED ISSUE #45717: Ansible tries to parse contents of diffs
I've tried to search for this problem in the issue tracker with the keywords `--diff`, `crash`, and `parse`, but haven't found anything.

##### SUMMARY

Diff printing seems to attempt to resolve variables in diffed files/parse them.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

Diff mode and `copy` module.

##### ANSIBLE VERSION

```
ansible 2.6.2                                                                                          
  config file = /home/laurens/repos/channable/devops/ansible/ansible.cfg
  configured module search path = ['/home/laurens/repos/channable/devops/ansible/modules']
  ansible python module location = /home/laurens/env/devops/lib/python3.6/site-packages/ansible
  executable location = /home/laurens/env/devops/bin/ansible
  python version = 3.6.5 (default, Apr  1 2018, 05:46:30) [GCC 7.3.0]
```

and

```
ansible 2.8.0.dev0
  config file = None
  configured module search path = ['/home/laurens/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/laurens/env/devops/lib/python3.6/site-packages/ansible
  executable location = /home/laurens/env/devops/bin/ansible
  python version = 3.6.5 (default, Apr  1 2018, 05:46:30) [GCC 7.3.0]
```

##### CONFIGURATION

No output. Running with default config.

```
```

##### OS / ENVIRONMENT

Originally seen in a more complex setting, but reproducible when running on Ubuntu 18.04, deploying with `ansible_connection=local`.

##### STEPS TO REPRODUCE

I've created a git repository which has a minimal test case: https://github.com/duijf/ansible-bug-report

Please tell me if anything is unclear.

##### EXPECTED RESULTS

Ansible does not crash and reports changed results correctly.

##### ACTUAL RESULTS

Ansible crashes.

```paste below
~/repos/channable/ansible-bug-report master
devops ❯ ansible-playbook -i inventory playbook.yml --diff

PLAY [localhost] ***************************************************************************************

TASK [Gathering Facts] *********************************************************************************
ok: [localhost]

TASK [problem : copy problematic file] *****************************************************************
--- before
+++ after: /home/laurens/repos/channable/ansible-bug-report/roles/problem/files/problem.yml
@@ -0,0 +1,7 @@
+The problem was originally found when copying prometheus
+templates, which can contain things like {{ $labels.instance }}.
+
+If this were a file in the `templates/` directory, I'd agree
+that this is asking for problems. However, this only gives
+problems when registering things to a variable and acting on
+it.                                                                                                    
                                                                                                        
changed: [localhost]                                                                                    
                                                                                                        
TASK [problem : inspect results] ***********************************************************************
fatal: [localhost]: FAILED! => {"msg": "An unhandled exception occurred while templating '{'diff': [{'before': '', 'after_header': '/home/laurens/repos/channable/ansible-bug-report/roles/problem/files/problem.yml', 'after': \"The problem was originally found when copying prometheus\\ntemplates, which can contain things like {{ $labels.instance }}.\\n\\nIf this were a file in the `templates/` directory, I'd agree\\nthat this is asking for problems. However, this only gives\\nproblems when registering things to a variable and acting on\\nit.\\n\"}], 'src': '/home/laurens/.ansible/tmp/ansible-tmp-1537170980.7106235-254323551122550/source', 'changed': True, 'group': 'laurens', 'uid': 1000, 'dest': '/tmp/problem.yml', 'checksum': '42a9a3edeb77bac53775ea092fab3cad962c0872', 'md5sum': '6e6ce6086486a66f1cdab91b98d0b7f0', 'owner': 'laurens', 'state': 'file', 'gid': 1000, 'mode': '0664', 'size': 308, 'failed': False}'. Error was a <class 'ansible.errors.AnsibleError'>, original message: template error while templating string: unexpected char '$' at 101. String: The problem was originally found when copying prometheus\ntemplates, which can contain things like {{ $labels.instance }}.\n\nIf this were a file in the `templates/` directory, I'd agree\nthat this is asking for problems. However, this only gives\nproblems when registering things to a variable and acting on\nit.\n"}                                                                       
        to retry, use: --limit @/home/laurens/repos/channable/ansible-bug-report/playbook.retry         
                                                                                                        
PLAY RECAP *********************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=1   
```

This only happens when:

 - `--diff` is used.
 - The file is changed, so the diff contains the variables.

It seems like Ansible is trying to interpret the output of the diff, since it prints `unexpected char '$' at 101`.

### PR DESCRIPTION
**Pull request by sivel**
##### SUMMARY
Ensure that the src file contents is converted to unicode in diff info. Fixes #45717

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
`lib/ansible/plugins/action/__init__.py`

##### ANSIBLE VERSION

```paste below
2.8
2.7
2.6
2.5
```

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Review by bcoca — APPROVED** [review]

**Comment by sivel** [comment]
@abadger I think you are correct about the tests being wrong.  Using the `difflist` from the tests produces:

```diff
--- before: just before
+++ after: just after
@@ -1 +1 @@
-['preface\nThe Before String\npostscript']
\ No newline at end of file
+['preface\nThe After String\npostscript']
\ No newline at end of file

--- before
+++ after
@@ -1 +1 @@
-['preface\nThe Before String\npostscript']
\ No newline at end of file
+['preface\nThe After String\npostscript']
\ No newline at end of file

diff skipped: source file appears to be binary
diff skipped: destination file appears to be binary
diff skipped: destination file size is greater than 1
diff skipped: source file size is greater than 2
what does prepared do?
```

So nothing was being done with the lists before, other than making them strings.

**Comment by abadger** [comment]
Cool.  This is ready for your review now, then.

File: lib/ansible/plugins/action/__init__.py [hunk 1]
```diff
@@ -1,3 +1,4 @@
+# coding: utf-8
```
Comment (webknjaz) [inline lib/ansible/plugins/action/__init__.py:1]: Nitpick: I know, that the codec check is flexible to recognize this, but it is more common to see `# -*- coding: utf-8 -*-` in the wild. So let's not confuse some future readers, who might be unaware of this.

**Comment by sivel** [comment]
Ok, this looks good to me.  I can't approve via review on my own PR :)

**Comment by sivel** [comment]
rebuild_merge

**Comment by sivel** [comment]
Ok, cool.  Seems I can't get CI to run on this PR.  I think the bot is cancelling it.

bot_status

**Pull request merged**